Expand `ConfigManager` tests

ddundo opened the pull request:
Hi @btobers, here is how I would expand what you did #78. Note that this branch merges into that branch :)

Here are the modifications I made:
- The types in `update_config` are compared against the original `/pygem/setup/config.yaml` file, since we know that everything there is correctly set up
- Types are more rigorously checked now. In particular, I had to specially update `ruamel.yaml.comments.CommentedSeq` types, which are, for example, `rgi_cols_drop` in `rgi`:
```
rgi:
  rgi_cols_drop:
    - GLIMSId
    - BgnDate
    - EndDate
    - Status
    - Linkages
    - Name
```
- I expanded tests that should now cover everything, including these types. 
 
You will see that one test fails, where I try to update an `int` value in `config.yaml` with a `float` value. This should be allowed.

If you agree with these changes, feel free to merge it into your branch. Could you also please try fixing this failing tests? You seem to be much more familiar with manipulating yaml files than me :)

btobers commented:
I like the idea of this, but I am now having a bunch of tests fail - seemingly because of TypeErrors.  The reason I was not as strict about enforcing type matching is because of the complexity of the yaml file.  The complicating factor is that some of the config key types in the source can be change. For example, the default (source) config has `climate.scenario = None`, but we want to allow users to set that key with a string of their choosing.  The same goes for `setup.glac_no`, which is also `None` by default, but can be set with a list of RGI6 id floats.  I am not sure how to handle these with your suggested changes.

ddundo commented:
Good point @btobers. I think we should explicitly handle this, with something like this:

```
import yaml

EXPECTED_TYPES = {
    "setup": dict,
    "setup.rgi_region01": list,
    "setup.rgi_region02": (list, str),
    "setup.glac_no_skip": (list, type(None)),
    "setup.glac_no": (list, type(None)),
    "setup.min_glac_area_km2": (int, float),
    "setup.include_landterm": bool,
    "setup.include_laketerm": bool,
    "setup.include_tidewater": bool,
    "setup.include_frontalablation": bool,

    "oggm": dict,
    "oggm.base_url": str,
    "oggm.logging_level": str,
    "oggm.border": int,
    "oggm.oggm_gdir_relpath": str,
    "oggm.overwrite_gdirs": bool,
    "oggm.has_internet": bool,
}

def validate_yaml(data, expected_types, prefix=""):
    for key, expected_type in expected_types.items():
        keys = key.split(".")
        sub_data = data
        for sub_key in keys:
            if sub_key in sub_data:
                sub_data = sub_data[sub_key]
            else:
                sub_data = None
                break

        if sub_data is not None and not isinstance(sub_data, expected_type):
            raise TypeError(f"Invalid type for '{key}': expected {expected_type}, got {type(sub_data)}")

with open("config.yaml", "r") as f:
    yaml_data = yaml.safe_load(f)

validate_yaml(yaml_data, EXPECTED_TYPES)
```

I only did it for a part of `config.yaml`, but we should extend it to cover everything. What do you think?

btobers commented:
Yeah, I think defining the type requirements would is a good idea.  I've got most of them started - just created a script to scrape the types form the  source config file.  Some of these need to be modified to allow multiple types (e.g. `setup.glac_no_skip` as you have).  Where should we put this validation?  I suppose if could go in `config.py` but maybe it's best to have these types defined in another module that gets sourced when `ConfigManager. validate_yaml()` is called?  For example, we could have this stored in a module such as `config_reqs.py`.

```
EXPECTED_TYPES = {
    "root": str,
    "user": dict,
    "user.name": str,
    "user.institution": str,
    "user.email": str,
    "setup": dict,
    "setup.rgi_region01": int,
    "setup.rgi_region02": str,
    "setup.glac_no_skip": NoneType,
    "setup.glac_no": NoneType,
    "setup.min_glac_area_km2": int,
    "setup.include_landterm": bool,
    "setup.include_laketerm": bool,
    "setup.include_tidewater": bool,
    "setup.include_frontalablation": bool,
    "oggm": dict,
    "oggm.base_url": str,
    "oggm.logging_level": str,
    "oggm.border": int,
    "oggm.oggm_gdir_relpath": str,
    "oggm.overwrite_gdirs": bool,
    "oggm.has_internet": bool,
    "climate": dict,
    "climate.ref_gcm_name": str,
    "climate.ref_startyear": int,
    "climate.ref_endyear": int,
    "climate.ref_wateryear": str,
    "climate.ref_spinupyears": int,
    "climate.gcm_name": str,
    "climate.scenario": NoneType,
    "climate.gcm_startyear": int,
    "climate.gcm_endyear": int,
    "climate.gcm_wateryear": str,
    "climate.constantarea_years": int,
    "climate.gcm_spinupyears": int,
    "climate.hindcast": bool,
    "climate.paths": dict,
    "climate.paths.era5_relpath": str,
    "climate.paths.era5_temp_fn": str,
    "climate.paths.era5_tempstd_fn": str,
    "climate.paths.era5_prec_fn": str,
    "climate.paths.era5_elev_fn": str,
    "climate.paths.era5_pressureleveltemp_fn": str,
    "climate.paths.era5_lr_fn": str,
    "climate.paths.cmip5_relpath": str,
    "climate.paths.cmip5_fp_var_ending": str,
    "climate.paths.cmip5_fp_fx_ending": str,
    "climate.paths.cmip6_relpath": str,
    "climate.paths.cesm2_relpath": str,
    "climate.paths.cesm2_fp_var_ending": str,
    "climate.paths.cesm2_fp_fx_ending": str,
    "climate.paths.gfdl_relpath": str,
    "climate.paths.gfdl_fp_var_ending": str,
    "climate.paths.gfdl_fp_fx_ending": str,
    "calib": dict,
    "calib.option_calibration": str,
    "calib.priors_reg_fn": str,
    "calib.HH2015_params": dict,
    "calib.HH2015_params.tbias_init": int,
    "calib.HH2015_params.tbias_step": int,
    "calib.HH2015_params.kp_init": float,
    "calib.HH2015_params.kp_bndlow": float,
    "calib.HH2015_params.kp_bndhigh": int,
    "calib.HH2015_params.ddfsnow_init": float,
    "calib.HH2015_params.ddfsnow_bndlow": float,
    "calib.HH2015_params.ddfsnow_bndhigh": float,
    "calib.HH2015mod_params": dict,
    "calib.HH2015mod_params.tbias_init": int,
    "calib.HH2015mod_params.tbias_step": float,
    "calib.HH2015mod_params.kp_init": int,
    "calib.HH2015mod_params.kp_bndlow": float,
    "calib.HH2015mod_params.kp_bndhigh": int,
    "calib.HH2015mod_params.ddfsnow_init": float,
    "calib.HH2015mod_params.method_opt": str,
    "calib.HH2015mod_params.params2opt": str,
    "calib.HH2015mod_params.ftol_opt": str,
    "calib.HH2015mod_params.eps_opt": float,
    "calib.emulator_params": dict,
    "calib.emulator_params.emulator_sims": int,
    "calib.emulator_params.overwrite_em_sims": bool,
    "calib.emulator_params.opt_hh2015_mod": bool,
    "calib.emulator_params.tbias_step": float,
    "calib.emulator_params.tbias_init": int,
    "calib.emulator_params.kp_init": int,
    "calib.emulator_params.kp_bndlow": float,
    "calib.emulator_params.kp_bndhigh": int,
    "calib.emulator_params.ddfsnow_init": float,
    "calib.emulator_params.option_areaconstant": bool,
    "calib.emulator_params.tbias_disttype": str,
    "calib.emulator_params.tbias_sigma": int,
    "calib.emulator_params.kp_gamma_alpha": int,
    "calib.emulator_params.kp_gamma_beta": int,
    "calib.emulator_params.ddfsnow_disttype": str,
    "calib.emulator_params.ddfsnow_mu": float,
    "calib.emulator_params.ddfsnow_sigma": float,
    "calib.emulator_params.ddfsnow_bndlow": int,
    "calib.emulator_params.ddfsnow_bndhigh": float,
    "calib.emulator_params.method_opt": str,
    "calib.emulator_params.params2opt": str,
    "calib.emulator_params.ftol_opt": str,
    "calib.emulator_params.eps_opt": float,
    "calib.MCMC_params": dict,
    "calib.MCMC_params.option_use_emulator": bool,
    "calib.MCMC_params.emulator_sims": int,
    "calib.MCMC_params.tbias_step": float,
    "calib.MCMC_params.tbias_stepsmall": float,
    "calib.MCMC_params.option_areaconstant": bool,
    "calib.MCMC_params.mcmc_step": float,
    "calib.MCMC_params.n_chains": int,
    "calib.MCMC_params.mcmc_sample_no": int,
    "calib.MCMC_params.mcmc_burn_pct": int,
    "calib.MCMC_params.thin_interval": int,
    "calib.MCMC_params.ddfsnow_disttype": str,
    "calib.MCMC_params.ddfsnow_mu": float,
    "calib.MCMC_params.ddfsnow_sigma": float,
    "calib.MCMC_params.ddfsnow_bndlow": int,
    "calib.MCMC_params.ddfsnow_bndhigh": float,
    "calib.MCMC_params.kp_disttype": str,
    "calib.MCMC_params.tbias_disttype": str,
    "calib.MCMC_params.tbias_mu": int,
    "calib.MCMC_params.tbias_sigma": int,
    "calib.MCMC_params.tbias_bndlow": int,
    "calib.MCMC_params.tbias_bndhigh": int,
    "calib.MCMC_params.kp_gamma_alpha": int,
    "calib.MCMC_params.kp_gamma_beta": int,
    "calib.MCMC_params.kp_lognorm_mu": int,
    "calib.MCMC_params.kp_lognorm_tau": int,
    "calib.MCMC_params.kp_mu": int,
    "calib.MCMC_params.kp_sigma": float,
    "calib.MCMC_params.kp_bndlow": float,
    "calib.MCMC_params.kp_bndhigh": float,
    "calib.data": dict,
    "calib.data.massbalance": dict,
    "calib.data.massbalance.hugonnet2021_relpath": str,
    "calib.data.massbalance.hugonnet2021_fn": str,
    "calib.data.massbalance.hugonnet2021_facorrected_fn": str,
    "calib.data.oib": dict,
    "calib.data.oib.oib_relpath": str,
    "calib.data.oib.oib_rebin": int,
    "calib.data.oib.oib_filter_pctl": int,
    "calib.data.frontalablation": dict,
    "calib.data.frontalablation.frontalablation_relpath": str,
    "calib.data.frontalablation.frontalablation_cal_fn": str,
    "calib.data.icethickness": dict,
    "calib.data.icethickness.h_consensus_relpath": str,
    "calib.icethickness_cal_frac_byarea": float,
    "sim": dict,
    "sim.option_dynamics": NoneType,
    "sim.option_bias_adjustment": int,
    "sim.nsims": int,
    "sim.out": dict,
    "sim.out.sim_stats": str,
    "sim.out.export_all_simiters": bool,
    "sim.out.export_extra_vars": bool,
    "sim.out.export_binned_data": bool,
    "sim.out.export_binned_components": bool,
    "sim.out.export_binned_area_threshold": int,
    "sim.oggm_dynamics": dict,
    "sim.oggm_dynamics.cfl_number": float,
    "sim.oggm_dynamics.cfl_number_calving": float,
    "sim.oggm_dynamics.glena_reg_relpath": str,
    "sim.oggm_dynamics.use_reg_glena": bool,
    "sim.oggm_dynamics.fs": int,
    "sim.oggm_dynamics.glen_a_multiplier": int,
    "sim.icethickness_advancethreshold": int,
    "sim.terminus_percentage": int,
    "sim.params": dict,
    "sim.params.use_constant_lapserate": bool,
    "sim.params.kp": int,
    "sim.params.tbias": int,
    "sim.params.ddfsnow": float,
    "sim.params.ddfsnow_iceratio": float,
    "sim.params.precgrad": float,
    "sim.params.lapserate": float,
    "sim.params.tsnow_threshold": int,
    "sim.params.calving_k": float,
    "mb": dict,
    "mb.option_surfacetype_initial": int,
    "mb.include_firn": bool,
    "mb.include_debris": bool,
    "mb.debris_relpath": str,
    "mb.option_elev_ref_downscale": str,
    "mb.option_temp2bins": int,
    "mb.option_adjusttemp_surfelev": int,
    "mb.option_prec2bins": int,
    "mb.option_preclimit": int,
    "mb.option_accumulation": int,
    "mb.option_ablation": int,
    "mb.option_ddf_firn": int,
    "mb.option_refreezing": str,
    "mb.Woodard_rf_opts": dict,
    "mb.Woodard_rf_opts.rf_month": int,
    "mb.HH2015_rf_opts": dict,
    "mb.HH2015_rf_opts.rf_layers": int,
    "mb.HH2015_rf_opts.rf_dz": int,
    "mb.HH2015_rf_opts.rf_dsc": int,
    "mb.HH2015_rf_opts.rf_meltcrit": float,
    "mb.HH2015_rf_opts.pp": float,
    "mb.HH2015_rf_opts.rf_dens_top": int,
    "mb.HH2015_rf_opts.rf_dens_bot": int,
    "mb.HH2015_rf_opts.option_rf_limit_meltsnow": int,
    "rgi": dict,
    "rgi.rgi_relpath": str,
    "rgi.rgi_lat_colname": str,
    "rgi.rgi_lon_colname": str,
    "rgi.elev_colname": str,
    "rgi.indexname": str,
    "rgi.rgi_O1Id_colname": str,
    "rgi.rgi_glacno_float_colname": str,
    "rgi.rgi_cols_drop": str,
    "time": dict,
    "time.option_leapyear": int,
    "time.startmonthday": str,
    "time.endmonthday": str,
    "time.wateryear_month_start": int,
    "time.winter_month_start": int,
    "time.summer_month_start": int,
    "time.option_dates": int,
    "time.timestep": str,
    "constants": dict,
    "constants.density_ice": int,
    "constants.density_water": int,
    "constants.area_ocean": float,
    "constants.k_ice": float,
    "constants.k_air": float,
    "constants.ch_ice": int,
    "constants.ch_air": int,
    "constants.Lh_rf": int,
    "constants.tolerance": float,
    "constants.gravity": float,
    "constants.pressure_std": int,
    "constants.temp_std": float,
    "constants.R_gas": float,
    "constants.molarmass_air": float,
    "debug": dict,
    "debug.refreeze": bool,
    "debug.mb": bool,
}
```

ddundo commented:
Nice! I think it's better to keep it in `ConfigManager`, since it's directly related.

btobers commented:
Fine with me.  I can create these two dictionaries - `EXPECTED_TYPES` AND `LIST_ELEMENT_TYPES` and push that to this branch.

btobers commented:
@ddundo I added all keys and list element types to the ConfigManager class and updated to call `validate_config` each time we `read_config` or `update_config` (before writing the udpates to file.  Would you be able to update the test_config functions you wrote as appropriate, and then we can merge this back to 72_update_config?

ddundo commented:
> @ddundo I added all keys and list element types to the ConfigManager class and updated to call `validate_config` each time we `read_config` or `update_config` (before writing the udpates to file. Would you be able to update the test_config functions you wrote as appropriate, and then we can merge this back to 72_update_config?

Thanks! And sure, I'll do that now

ddundo commented on pygem/setup/config.py:
```diff
@@ -24,101 +24,319 @@ def __init__(self, config_filename='config.yaml', base_dir=None, overwrite=False
-    def create_config(self):
+    def copy_source_config(self):
         """Copy the default configuration file to the expected location"""
-        if not os.path.exists(self.source_config_path):
-            raise FileNotFoundError(f"Default config file not found at {self.source_config_path}, there may have been an installation issue")
```
I removed this because the config file `pygem/setup/config.yaml` should always exist

ddundo commented on pygem/setup/config.py:
```diff
@@ -24,101 +24,319 @@ def __init__(self, config_filename='config.yaml', base_dir=None, overwrite=False
+            user_config = ryaml.load(f)  # Using ruamel.yaml for preservation
 
         if validate:
-            self.compare_with_source()
+            self.validate_config(user_config)
```
I think we should always validate

ddundo commented on pygem/tests/test_config.py:
```diff
@@ -1,115 +1,129 @@
+        assert config["user"]["email"] == "updated@example.com"
+        assert config["rgi"]["rgi_cols_drop"] == ["Item1", "Item2"]
+
+    def test_update_config_dict(self):
+        """Test that update_config updates the config file for nested dictionaries."""
+        # Values before updating
+        config = self.config_manager.read_config()
+        assert config["user"]["name"] == "David Rounce"
+        assert config["user"]["email"] == "drounce@cmu.edu"
+        assert config["user"]["institution"] == "Carnegie Mellon University, Pittsburgh PA"
+
+        updates = {
+            "user": {
+                "name": "New Name",
+                "email": "New email",
+                "institution": "New Institution",
+            }
+        }
+        self.config_manager.update_config(updates)
+
+        # Values after updating
+        config = self.config_manager.read_config()
+        assert config["user"]["name"] == "New Name"
+        assert config["user"]["email"] == "New email"
+        assert config["user"]["institution"] == "New Institution"
```
This will fail if one of "name", "email, and "institution" isn't provided. For example, if you don't provide email: `KeyError: 'Missing required key in configuration: user.email'`

So it would be good to fix this if you have time. Otherwise could you please open an issue describing this so we can fix it in the future?

btobers replied:
I set each of these as  (str, type(None)) in `EXPECTED_TYPES`, so they can now be left empty.

btobers reviewed (approved):
All tests passed!
